src/common/scrub_types.cc

```diff
@@ -74,16 +74,15 @@ void shard_info_wrapper::encode(bufferlist& bl) const
   ENCODE_START(3, 3, bl);
   encode(errors, bl);
   encode(primary, bl);
-  if (has_shard_missing()) {
```

Comment:
@jdurgin Old clients when `has_shard_missing` skip `DECODE_FINISH()` so they don't throw an exception if the decoding went past offset computed from `struct_len`.  In addition, they don't `bl.advance()` if the number of bytes decoded doesn't include all the encoded data.  Assuming `struct_len` was zero when coming from an old server no exception would be thrown.  Also, because we have a `vector` of buffer lists it doesn't matter if we don't advance() because we are done with the buffer list.

An older server will put in struct_v and I assume the `struct_compat` and `struct_len` are zero because `ENCODE_FINISH()` is skipped.  In a new client as before the `DECODE_START()` was just working because of the zero values.  It will compute `struct_end` as zero which causes a new client's `DECODE_FINISH()` to do nothing.

I can't think a scenario that would break anything including if the structure version changes and/or additional fields are added.

Reply:
cool, thanks for explaining the reasoning